xarray/core/datatree.py

```diff
@@ -158,6 +158,19 @@ def _check_alignment(
             _check_alignment(child_path, child_ds, base_ds, child.children)
 
 
+def _check_for_slashes_in_names(variables: Iterable[Hashable]) -> None:
+    offending_variable_names = [
+        name for name in variables if isinstance(name, str) and "/" in name
+    ]
+    if len(offending_variable_names) > 0:
+        raise KeyError(
+            "Given Dataset contains variable names with '/': "
+            f"{offending_variable_names}. "
+            "A Dataset represents a group, and a single group "
+            "cannot have path-like variable names with '/' characters in them. "
+        )
```

Comment:
I think this is good, but @keewis you mentioned you had a comment didn't you?

Reply:
I'm not sure if this has to be considered here, but NetCDF4 and HDF5 do not allow forward slashes in variable/dataset names.

Reply:
Interesting. I think that doesn't change anything for this PR, so long as that restriction is already explicitly accounted for in the respective backend save methods.